services/ens/api.go

```diff
@@ -0,0 +1,312 @@
+	return &owner, nil
+}
+
+func (api *API) ContentHash(ctx context.Context, chainID uint64, username string) ([]byte, error) {
+	err := validateENSUsername(username)
+	if err != nil {
+		return nil, err
+	}
+
+	resolverAddress, err := api.Resolver(ctx, chainID, username)
+	if err != nil {
+		return nil, err
+	}
+
+	resolver, err := api.contractMaker.newPublicResolver(chainID, resolverAddress)
+	if err != nil {
+		return nil, err
+	}
+
+	callOpts := &bind.CallOpts{Context: ctx, Pending: false}
+	contentHash, err := resolver.Contenthash(callOpts, nameHash(username))
+	if err != nil {
+		return nil, nil
+	}
+
+	return contentHash, nil
+}
+
+func (api *API) PublicKeyOf(ctx context.Context, chainID uint64, username string) ([32]byte, [32]byte, error) {
+	empty := [32]byte{}
+	err := validateENSUsername(username)
+	if err != nil {
+		return empty, empty, err
+	}
+
+	resolverAddress, err := api.Resolver(ctx, chainID, username)
+	if err != nil {
+		return empty, empty, err
+	}
+
+	resolver, err := api.contractMaker.newPublicResolver(chainID, resolverAddress)
+	if err != nil {
+		return empty, empty, err
+	}
+
+	callOpts := &bind.CallOpts{Context: ctx, Pending: false}
+	pubKey, err := resolver.Pubkey(callOpts, nameHash(username))
+	if err != nil {
+		return empty, empty, nil
+	}
+
+	return pubKey.X, pubKey.Y, nil
+}
+
+func (api *API) AddressOf(ctx context.Context, chainID uint64, username string) (*common.Address, error) {
+	err := validateENSUsername(username)
+	if err != nil {
+		return nil, err
+	}
+
+	resolverAddress, err := api.Resolver(ctx, chainID, username)
+	if err != nil {
+		return nil, err
+	}
+
+	resolver, err := api.contractMaker.newPublicResolver(chainID, resolverAddress)
+	if err != nil {
+		return nil, err
+	}
+
+	callOpts := &bind.CallOpts{Context: ctx, Pending: false}
+	addr, err := resolver.Addr(callOpts, nameHash(username))
+	if err != nil {
+		return nil, err
+	}
+
+	return &addr, nil
+}
+
+func (api *API) ExpireAt(ctx context.Context, chainID uint64, username string) (*big.Int, error) {
+	err := validateENSUsername(username)
+	if err != nil {
+		return nil, err
+	}
+
+	registrar, err := api.contractMaker.newUsernameRegistrar(chainID)
+	if err != nil {
+		return nil, err
+	}
+
+	callOpts := &bind.CallOpts{Context: ctx, Pending: false}
+	expTime, err := registrar.GetExpirationTime(callOpts, nameHash(username))
+	if err != nil {
+		return nil, err
+	}
+
+	return expTime, nil
+}
+
+func (api *API) Price(ctx context.Context, chainID uint64) (*big.Int, error) {
+	registrar, err := api.contractMaker.newUsernameRegistrar(chainID)
+	if err != nil {
+		return nil, err
+	}
+
+	callOpts := &bind.CallOpts{Context: ctx, Pending: false}
+	price, err := registrar.GetPrice(callOpts)
+	if err != nil {
+		return nil, err
+	}
+
+	return price, nil
+}
+
+func (api *API) Release(ctx context.Context, chainID uint64, from string, gasPrice *big.Int, gasLimit uint64, password string, username string) (string, error) {
+	err := validateENSUsername(username)
```

Comment:
It's probably a better idea to receive a `transactions.SendTxArgs` instead of  from `string, gasPrice *big.Int, gasLimit uint64, password string`. `SendTxArgs` also has supports for other trx fields like the Nonce, and EIP1559 MaxFeePerGas  and MaxPriorityFeePerGas

Reply:
If you use the transactions.SendTxArgs, you'd need a mapping function to convert from that struct to a `bind.TransactOpts`

Reply:
:+1: